programs/uxd/src/instructions/mint_with_router.rs

```diff
@@ -0,0 +1,430 @@
+                .redeemable_amount_under_management,
+            redeemable_amount_under_management_cap: mercurial_vault_depository_0
+                .redeemable_amount_under_management_cap,
+        },
+        // Credix Lp Depository 0 details
+        DepositoryInfoForMintWithRouter {
+            weight_bps: controller.credix_lp_depository_0_weight_bps,
+            redeemable_amount_under_management: credix_lp_depository_0
+                .redeemable_amount_under_management,
+            redeemable_amount_under_management_cap: credix_lp_depository_0
+                .redeemable_amount_under_management_cap,
+        },
+    ];
+
+    drop(controller);
+    drop(identity_depository);
+    drop(mercurial_vault_depository_0);
+    drop(credix_lp_depository_0);
+
+    // Compute the desired target amounts for each depository
+    let depositories_target_redeemable_amount = calculate_depositories_target_redeemable_amount(
+        redeemable_circulating_supply_after,
+        &depository_info
+            .iter()
+            .map(|depository_info| DepositoryInfoForTargetRedeemableAmount {
+                weight_bps: depository_info.weight_bps,
+                redeemable_amount_under_management_cap: depository_info
+                    .redeemable_amount_under_management_cap,
+            })
+            .collect::<Vec<DepositoryInfoForTargetRedeemableAmount>>(),
+    )?;
+
+    // Compute the desired mint amounts for each depository
+    let depositories_mint_collateral_amount = calculate_depositories_mint_collateral_amount(
+        collateral_amount,
+        &std::iter::zip(
+            depository_info.iter(),
+            depositories_target_redeemable_amount.iter(),
+        )
+        .map(|(depository_info, depository_target_redeemable_amount)| {
+            DepositoryInfoForMintCollateralAmount {
+                target_redeemable_amount: *depository_target_redeemable_amount,
+                redeemable_amount_under_management: depository_info
+                    .redeemable_amount_under_management,
+            }
+        })
+        .collect::<Vec<DepositoryInfoForMintCollateralAmount>>(),
+    )?;
+
+    // Mint the desired amount at identity_depository
+    let identity_depository_mint_collateral_amount =
+        depositories_mint_collateral_amount[ROUTER_IDENTITY_DEPOSITORY_INDEX];
+    msg!(
+        "[mint_with_router:mint_with_identity_depository:{}]",
+        identity_depository_mint_collateral_amount
+    );
+    if identity_depository_mint_collateral_amount > 0 {
+        uxd_cpi::cpi::mint_with_identity_depository(
+            ctx.accounts.into_mint_with_identity_depository_context(),
+            identity_depository_mint_collateral_amount,
+        )?;
+    }
+
+    // Mint the desired amount at mercurial_vault_depository_0
+    let mercurial_vault_depository_0_mint_collateral_amount =
+        depositories_mint_collateral_amount[ROUTER_MERCURIAL_VAULT_DEPOSITORY_0_INDEX];
```

Comment:
The trans-file indexing system here is likely to create mistakes.

**quick soluce**: use 0, 1, 2 index here right after `calculate_depositories_mint_collateral_amount` with a comment stating to respect the same order as  during ```let depository_info = vec![```.

**soluce**: embed `depositories_mint_collateral_amount` info into the `DepositoryInfoForMintWithRouter` structure and add an abstract function `mint()` to it using `Trait`. So you can just loop on the vector and call the `mint` cutom function defined for all type of depositories. No more mistakes possible.

Reply:
Looking into this :loading:

Reply:
Done here:
https://github.com/UXDProtocol/uxd-program/pull/246/commits/c6a4d0276e7f984c549c1b4f3c09d54c7f652f6d

Or myt interpretation of it at least :D